### DISCUSSION TITLE
CPP 回测没有买卖信号

### DISCUSSION THREAD

**Discussion by rafaelchan** [wondertrader/wondertrader#64] [想法/Ideas]
用dist下已经编译好的dll测试时有买卖信号的。
代码是dev分支最新版本，cta bt， WtCtaStraFact.dll 是demo目录下编译的。
执行结果如下：

![1651549222(1)](https://user-images.githubusercontent.com/37570139/166402190-90477c9c-e9c6-42f9-880b-92ec1fba96f1.png)

跟踪代码发现:
bool CtaMocker::on_schedule(uint32_t curDate, uint32_t curTime)
{
	_is_in_schedule = true;//开始调度,修改标记

	_schedule_times++;

	bool isMainUdt = false;
	bool emmited = false;

	for (auto it = _kline_tags.begin(); it != _kline_tags.end(); it++)
	{
		const std::string& key = it->first;
		KlineTag& marker = (KlineTag&)it->second;

		StringVector ay = StrUtil::split(key, "#");
		const char* stdCode = ay[0].c_str();

		if (key == _main_key)
		{
			if (marker._closed)
			{
				isMainUdt = true;
				marker._closed = false;
			}
			else
			{
				isMainUdt = false;     //
				break;
			}
		}
.....
isMainUdt  一直是false,跳出循环，没有触发策略的on_schedule。  请问这是什么原因导致的。

**Comment by rafaelchan** [comment]
在He!J的帮助下问题解决了。
marker._closed 是否闭合是在CtaMocker的on_bar 里设置的
{
     ...

	thread_local static char realPeriod[8] = { 0 };
	fmtutil::format_to(realPeriod, "{}{}", period[0], times);     // period[0]改成 period 或 &period[0]

	thread_local static char key[64] = { 0 };
	fmtutil::format_to(key, "{}#{}", stdCode, realPeriod);

	KlineTag& tag = _kline_tags[key];
	tag._closed = true;
        ...
}
原因：fmtutil::format_to(realPeriod, "{}{}", period[0], times)这行代码在某些编译环境下格式后是空值，这是最后拼成的key值和on_schedule（）的判断的key值就不一致了，导致on_schedule获取的marker._closed的值就一直是false。
解决方法：fmtutil::format_to(realPeriod, "{}{}", period[0], times);    中第二参数 period[0]改成 period 或 &period[0]。
PS：问题在有些编译环境下不会发生，这里可能就需要细究fmtutil::format_to 的实现和底层编译环境了。

**Reply by ZzzzHeJ** [reply]
good job

**Reply by wondertrader** [reply]
这个地方最新dev改掉了，这个原因我也解释不大清楚
